src/builder.rs

```diff
@@ -935,23 +935,32 @@ impl<P: fmt::Debug, S: InProgressSignatures> Authorization for InProgress<P, S>
 ///
 /// This struct contains the private data needed to create a [`Proof`] for a [`Bundle`].
 #[derive(Clone, Debug)]
-pub struct Unproven<C: OrchardCircuit> {
-    circuits: Vec<Circuit<C>>,
+pub struct Unproven {
+    witnesses: Vec<Witnesses>,
 }
 
-impl<S: InProgressSignatures, C: OrchardCircuit> InProgress<Unproven<C>, S> {
+impl<S: InProgressSignatures> InProgress<Unproven, S> {
     /// Creates the proof for this bundle.
-    pub fn create_proof(
+    pub fn create_proof<C: OrchardCircuit>(
         &self,
         pk: &ProvingKey,
         instances: &[Instance],
         rng: impl RngCore,
     ) -> Result<Proof, halo2_proofs::plonk::Error> {
-        Proof::create(pk, &self.proof.circuits, instances, rng)
+        let circuits: Vec<Circuit<C>> = self
+            .proof
+            .witnesses
+            .iter()
+            .map(|witnesses| Circuit::<C> {
+                witnesses: witnesses.clone(),
+                phantom: std::marker::PhantomData,
+            })
+            .collect();
+        Proof::create(pk, &circuits, instances, rng)
```

Comment:
Open question: 
Can we add some kind of protection from triggering ZSA proof generation using Vanilla witness or vice versa?
what is the cost?

Reply:
Added circuit_flavor into Unproven struct